agent/php_stacked_segment.h

```diff
@@ -155,6 +155,397 @@
+ * segments are cleaned when the next scenario 1-5 occurs.
+ *
+ * case 5 in newrelic APIs that depend on having the current segment
+ *  root <                      root                      root
+ *                               |                         |
+ *                               *A <                      *A
+ *                                                         |
+ *                                                         *B
+ *
+ * We start out with a root segment, OAPI calls nr_php_observer_fcall_begin(A),
+ * and it starts stacked segment *A and then nr_php_observer_fcall_begin(B)
+ * starts stacked segment *B as a child of *A.
+ *
+ *  root                       root
+ *   |                           |
+ *   *A                          *A
+ *   |                           |
+ *   *B                          *B <
+ *   |                           |
+ *   *C <                        c
+ *
+ * nr_php_observer_fcall_begin(C) starts *C gets started as child
+ * of *B. Function C throws an uncaught exception which B does not catch so
+ * netiher nr_php_observer_fcall_end(B) nor nr_php_observer_fcall_end(C) is
+ * called and *C remains the current segment. A catches the exception and calls
+ * newrelic_notice_error. We check the `this` value of the function that called
+ * newrelic_notice_error and see it is not the same. Because we received no
+ * nr_php_observer_fcall_end up to that point, we know the exception was
+ * uncaught until the Function A. We check the global exception hook and see it
+ * has a value and that the global uncaught_exception_this also matches the
+ * current segment `this`. Time to apply the exception and clean up dangling
+ * segments as we don't want to apply the notice_error to the wrong segment. We
+ * pop the current segment *C and apply the exception. Because it has an
+ * exception, the segment is kept so we copy the contents of the stacked segment
+ * *C into a segment c we obtained from the slab allocator, and we make c a
+ * child of the stacked segment *B which becomes the current segment.
+ *
+ *  root
+ *   |
+ *   *A <
+ *    |
+ *    b
+ *    |
+ *    c
+ *
+ *
+ * But we aren't done yet.
+ * The `this` value of the function that called
+ * newrelic_notice_error and see it is not the same as the current segment
+ * `this`. We pop the current segment *B and apply the exception. Because it has
+ * an exception, the segment is kept so we copy the contents of the stacked
+ * segment *B into a segment b we obtained from the slab allocator, and we make
+ * b a child of the stacked segment *A which becomes the current segment.  The
+ * this` value of the function that called newrelic_notice_error and see it is
+ * not the same as the current segment `this` so it proceeds and applies the
+ * notice error to the current segment *A.
+ * Note that this only works with segments on the default parent stack.
+ * Stacked segments cannot be used to model async segments.
+ */
+
```

Comment:
This whole writeup/comment block is really great! Thanks for putting all these detailed cases together and using ASCII diagrams to demonstrate what is happening. Everything is explained very clearly.